superduperdb/components/graph.py

```diff
@@ -0,0 +1,220 @@
+import typing as t
+from collections import OrderedDict
+
+import networkx as nx
+
+from superduperdb.components.component import Component
+from superduperdb.components.model import Model, _Predictor
+
+
+class GraphModel:
+    def __init__(self, name):
+        self.identifier = name
+        self.inputs = []
+
+    def predict(self, x):
+        return x
+
+
+class IntermediateGraphNode:
+    def __init__(self, node, graph):
+        self.node = node
+        self.G = graph
+
+    @property
+    def output(self):
+        return self.G.get_output(self.node)
+
+
+@dc.dataclass(kw_only=True)
+class Graph(Component, _Predictor):
+    models: t.List[Model] = dc.field(default_factory=list)
+    edges: t.List[t.Tuple[str, str, t.Union[None, str]]] = dc.field(
+        default_factory=list
+    )
+    _DEFAULT_ARG_WEIGHT: t.ClassVar[str] = '_base'
+    type_id: t.ClassVar[str] = 'graph'
+
+    def __post_init__(self, artifacts):
+        self.G = nx.DiGraph()
+        self.nodes = {}
+        self._node_output_cache = {}
+        self.version = 0
+        self._db = None
+
+        # Load the models and edges into a di graph
+        models = {m.identifier: m for m in self.models}
+        if self.edges and models:
+            for connection in self.edges:
+                u, v, on = connection
+                self.connect(
+                    models[u] if u != self.identifier else self,
+                    models[v],
+                    on=on,
+                    update_edge=False,
+                )
+        super().__post_init__(artifacts=artifacts)
+
+    def connect(
+        self,
+        u: Component,
+        v: Model,
+        on: t.Optional[str] = None,
+        update_edge: t.Optional[bool] = True,
+    ):
+        assert isinstance(u, (Model, Graph))
+        assert isinstance(v, Model)
+
+        if u.identifier not in self.nodes:
+            if isinstance(u, Graph):
+                self.nodes[u.identifier] = GraphModel(self.identifier)
+            else:
+                self.nodes[u.identifier] = u
+            self.G.add_node(u.identifier)
+
+        if v.identifier not in self.nodes:
+            self.nodes[v.identifier] = v
+            self.G.add_node(v.identifier)
+
+        G_ = self.G.copy()
+        G_.add_edge(u.identifier, v.identifier, weight=on or self._DEFAULT_ARG_WEIGHT)
+
+        if not nx.is_directed_acyclic_graph(G_):
+            raise TypeError('The graph is not DAG with this edge')
+        self.G = G_
+
+        if update_edge:
+            self.edges.append(
+                (u.identifier, v.identifier, on or self._DEFAULT_ARG_WEIGHT)
+            )
+            if isinstance(u, Model) and u not in self.models:
+                self.models.append(u)
+            if v not in self.models:
+                self.models.append(v)
+        return IntermediateGraphNode(self.nodes[v.identifier], self)
+
+    def stash_node_output(self, node, output):
+        self._node_output_cache[node.identifier] = output
```

Comment:
I think we need to create a separate variable `_node_output_cache` for each prediction.